webknossos/webknossos/skeleton/graph.py

```diff
@@ -13,51 +13,143 @@
+        return len(self._id_to_attrs)
+
+    def get_node(self, id_: int) -> Node:
+        return self._id_to_node[id_]
+
+
+class _AdjDict(MutableMapping):
+    """Dict-like container for the `Graph` class below to handle both nodes and ids as keys.
+    Needs a reference to the _node attribute (of class `_NodeDict`) of the graph object
+    to get a reference from ids to nodes. Iterating over the keys yields the `Node` objects."""
+
+    def __init__(self, *, node_dict: _NodeDict) -> None:
+        self._id_to_attrs: Dict[int, Any] = {}
+        self._node_dict = node_dict
+
+    def __getitem__(self, key: Union[Node, int]) -> Any:
+        return self._id_to_attrs[_get_id(key)]
+
+    def __setitem__(self, key: Node, value: Dict) -> None:
+        self._id_to_attrs[_get_id(key)] = value
+
+    def __delitem__(self, key: Union[Node, int]) -> None:
+        del self._id_to_attrs[_get_id(key)]
+
+    def __iter__(self) -> Iterator[Node]:
+        return (self._node_dict.get_node(i) for i in self._id_to_attrs)
+
+    def __len__(self) -> int:
+        return len(self._id_to_attrs)
+
+
+class Graph(nx.Graph):
     """
     Contains a collection of nodes and edges.
+    This class inherits from [`networkx.Graph`](https://networkx.org/documentation/stable/reference/classes/graph.html).
+    For further methods, please [check the networkx documentation](https://networkx.org/documentation/stable/reference/classes/graph.html#methods).
     """
 
-    name: str
-    group: "Group" = attr.ib(eq=False, repr=False)
-    _nml: "Skeleton" = attr.ib(eq=False, repr=False)
-    _id: int = attr.ib(init=False)
-    nx_graph: nx.Graph = attr.ib(
-        init=False,
-        repr=lambda nx_graph: f"<n={len(nx_graph.nodes)},e={len(nx_graph.edges)}>",
-        eq=lambda nx_graph: (
-            sorted(nx_graph.nodes(data="obj")),
-            sorted(nx_graph.edges),
-        ),
-    )
-    color: Optional[Vector4] = None
-    _enforced_id: Optional[int] = attr.ib(None, eq=False, repr=False)
-
-    def __attrs_post_init__(self) -> None:
-        self.nx_graph = nx.Graph()
-        if self._enforced_id is not None:
-            self._id = self._enforced_id
+    def __init__(
+        self,
+        name: str,
+        group: "Group",
+        skeleton: "Skeleton",
+        color: Optional[Vector4] = None,
+        enforced_id: Optional[int] = None,
+    ) -> None:
+        # To be able to reference nodes by id after adding them for the first time, we use custom dict-like classes
+        # for the networkx-graph structures, that have nodes as keys:
+        # * `self._node`: _NodeDict
+        #   holding the attributes of nodes, keeping references from ids to nodes
+        # * `self._adj`: _AdjDict on the first two levels
+        #   holding edge attributes on the last level, using self._node to convert ids to nodes
+        #
+        # For further details, see the *Subclasses* section here: https://networkx.org/documentation/stable/reference/classes/graph.html
```

Comment:
> the docs state "To replace one of the dicts create a new graph class by changing the class(!) variable holding". however, you are setting the instance variable. probably, because _AdjDict needs the reference to _node?

Exactly :+1: 

> Do you think, this is fine? I'm just wondering because the docs explicitly highlighted this with the exclamation mark.

Yes, since we set this up *before* `nx.Graph.__init__()` is called. I guess the exclamation mark is there to emphasize that it doesn't help to set it on the instance after it's creation (e.g. `g = nx.Graph; g.node_dict_factory = …`). But doing this before the parents init is totally fine, since the attribute on the instance shadows the class-attribute. I'll add an explaining comment in our Graph's `__init__`.

> node_dict_factory could at least be turned into a class variable probably?

Yes, this is possible. But I don't see how this would help, and I'd find this even more confusing.

Reply:
Ok, I agree :+1: